File: boomega-gui/src/main/java/com/dansoftware/boomega/gui/api/EmptyContext.java
/*
 * Boomega - A modern book explorer & catalog application
 * Copyright (C) 2020-2022  Daniel Gyoerffy
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.dansoftware.boomega.gui.api;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.EventHandler;
import javafx.scene.Node;
import javafx.scene.Scene;
import javafx.scene.control.ButtonType;
import javafx.scene.control.Hyperlink;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.Region;
import javafx.stage.Window;
import javafx.util.Duration;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.function.Consumer;

public interface EmptyContext extends Context {
    @Override
    default void showOverlay(Region region, boolean blocking) {

    }

    @Override
    default void hideOverlay(Region region) {

    }

    @Override
    @NotNull
    default ContextDialog showErrorDialog(String title, String message, Consumer<ButtonType> onResult) {
        return null;
    }

    @Override
    @NotNull
    default ContextDialog showErrorDialog(String title, String message, Exception cause, Consumer<ButtonType> onResult) {
        return null;
    }

    @Override
    @NotNull
    default ContextDialog showInformationDialog(String title, String message, Consumer<ButtonType> onResult) {
        return null;
    }

    @Override
    default ContextDialog showConfirmationDialog(String title, String message, Consumer<ButtonType> onResult) {
        return null;
    }

    @Override
    @NotNull
    default ContextDialog showDialog(String title, Node content, Consumer<ButtonType> onResult, ButtonType... buttonTypes) {
        return null;
    }

    @Override
    default ButtonType showErrorDialogAndWait(String title, String message) {
        return null;
    }

    @Override
    default ButtonType showErrorDialogAndWait(String title, String message, Exception e) {
        return null;
    }

    @Override
    default ButtonType showInformationDialogAndWait(String title, String message) {
        return null;
    }

    @Override
    default ButtonType showConfirmationDialogAndWait(String title, String message) {
        return null;
    }

    @Override
    default ButtonType showDialogAndWait(String title, Node content, ButtonType... buttonTypes) {
        return null;
    }

    @Override
    default void showErrorNotification(String title, String message) {

    }

    @Override
    default void showErrorNotification(String title, String message, EventHandler<MouseEvent> onClicked) {

    }

    @Override
    default void showErrorNotification(String title, String message, Duration duration) {

    }

    @Override
    default void showErrorNotification(String title, String message, Duration duration, EventHandler<MouseEvent> onClicked) {

    }

    @Override
    default void showWarningNotification(String title, String message) {

    }

    @Override
    default void showWarningNotification(String title, String message, EventHandler<MouseEvent> onClicked) {

    }

    @Override
    default void showWarningNotification(String title, String message, Duration duration) {

    }

    @Override
    default void showWarningNotification(String title, String message, Duration duration, EventHandler<MouseEvent> onClicked) {

    }

    @Override
    default void showInformationNotification(String title, String message) {

    }

    @Override
    default void showInformationNotification(String title, String message, EventHandler<MouseEvent> onClicked) {

    }

    @Override
    default void showInformationNotification(String title, String message, Duration duration) {

    }

    @Override
    default void showInformationNotification(String title, String message, Duration duration, EventHandler<MouseEvent> onClicked) {

    }

    @Override
    default void showInformationNotification(String title, String message, EventHandler<MouseEvent> onClicked, Hyperlink... hyperlinks) {

    }

    @Override
    @Nullable
    default Scene getContextScene() {
        return null;
    }

    @Override
    @Nullable
    default Window getContextWindow() {
        return null;
    }

    @Override
    default void focusRequest() {

    }

    @Override
    default void toFrontRequest() {

    }

    @Override
    default boolean isShowing() {
        return false;
    }

    @Override
    default void showIndeterminateProgress() {

    }

    @Override
    default void stopProgress() {

    }

    @Override
    default void showProgress(long done, long max, @NotNull ProgressType type) {

    }

    @Override
    default void onWindowPresent(Consumer<Window> action) {

    }

    @NotNull
    @Override
    default ObservableList<Region> getBlockingOverlaysShown() {
        return FXCollections.observableArrayList();
    }

    @NotNull
    @Override
    default ObservableList<Region> getNonBlockingOverlaysShown() {
        return FXCollections.observableArrayList();
    }

    @Override
    default void showOverlay(@NotNull Region region) {
        Context.super.showOverlay(region);
    }

    @Override
    default void showErrorDialog(@NotNull String title, @NotNull String message) {
        Context.super.showErrorDialog(title, message);
    }

    @Override
    default void showErrorDialog(@NotNull String title, @NotNull String message, @Nullable Exception e) {
        Context.super.showErrorDialog(title, message, e);
    }

    @Override
    default void sendRequest(@NotNull Context.Request request) {
        Context.super.sendRequest(request);
    }

    @Override
    default boolean isReachable() {
        return Context.super.isReachable();
    }

    @Override
    default void close() {
        Context.super.close();
    }
}

File: src/main/java/com/dansoftware/libraryapp/gui/context/EmptyContext.kt
package com.dansoftware.libraryapp.gui.context

import javafx.event.EventHandler
import javafx.scene.Node
import javafx.scene.Scene
import javafx.scene.control.ButtonType
import javafx.scene.input.MouseEvent
import javafx.scene.layout.Region
import javafx.stage.Window
import javafx.util.Duration
import java.util.function.Consumer

private class EmptyContext : Context {

    override fun showOverlay(region: Region?, blocking: Boolean) {}

    override fun hideOverlay(region: Region?) {}

    override fun showErrorDialog(
        title: String?,
        message: String?,
        onResult: Consumer<ButtonType>?
    ): ContextDialog = null!!

    override fun showErrorDialog(
        title: String?,
        message: String?,
        exception: Exception?,
        onResult: Consumer<ButtonType>?
    ): ContextDialog = null!!

    override fun showInformationDialog(
        title: String?,
        message: String?,
        onResult: Consumer<ButtonType>?
    ): ContextDialog = null!!

    override fun showConfirmationDialog(
        title: String?,
        message: String?,
        onResult: Consumer<ButtonType>?
    ): ContextDialog = null!!

    override fun showDialog(
        title: String?,
        content: Node?,
        onResult: Consumer<ButtonType>?,
        vararg buttonTypes: ButtonType?
    ): ContextDialog = null!!

    override fun showErrorDialogAndWait(title: String?, message: String?): ButtonType? = null

    override fun showErrorDialogAndWait(title: String?, message: String?, e: java.lang.Exception?): ButtonType? = null

    override fun showInformationDialogAndWait(title: String?, message: String?): ButtonType? = null

    override fun showConfirmationDialogAndWait(title: String?, message: String?): ButtonType? = null

    override fun showDialogAndWait(title: String?, content: Node?, vararg buttonTypes: ButtonType?): ButtonType? = null

    override fun showErrorNotification(title: String?, message: String?) {}

    override fun showErrorNotification(title: String?, message: String?, onClicked: EventHandler<MouseEvent>?) {}

    override fun showErrorNotification(title: String?, message: String?, duration: Duration?) {}

    override fun showErrorNotification(
        title: String?,
        message: String?,
        duration: Duration?,
        onClicked: EventHandler<MouseEvent>?
    ) {}

    override fun showWarningNotification(title: String?, message: String?) {}

    override fun showWarningNotification(title: String?, message: String?, onClicked: EventHandler<MouseEvent>?) {}

    override fun showWarningNotification(title: String?, message: String?, duration: Duration?) {}

    override fun showWarningNotification(
        title: String?,
        message: String?,
        duration: Duration?,
        onClicked: EventHandler<MouseEvent>?
    ) {}

    override fun showInformationNotification(title: String?, message: String?) {}

    override fun showInformationNotification(title: String?, message: String?, onClicked: EventHandler<MouseEvent>?) {}

    override fun showInformationNotification(title: String?, message: String?, duration: Duration?) {}

    override fun showInformationNotification(
        title: String?,
        message: String?,
        duration: Duration?,
        onClicked: EventHandler<MouseEvent>?
    ) {}

    override fun showModule(classRef: Class<*>) {}

    override fun <D> showModule(classRef: Class<out NotifiableModule<D>>, data: D) {}

    override fun getContextScene(): Scene? = null

    override fun getContextWindow(): Window? = null

    override fun requestFocus() {}

    override fun toFront() {}

    override fun isShowing(): Boolean = false

    override fun showIndeterminateProgress() {}

    override fun stopProgress() {}

    override fun showProgress(done: Long, max: Long, type: Context.ProgressType) {}

    override fun onWindowPresent(action: Consumer<Window>?) {}

    override fun isReachable(): Boolean = false
}